### PR TITLE
#185 #186 #187: Handling of multipleReferencesAllowed and arrays

### PR DESCRIPTION
**Pull request by reckart**
#185 - Transitively referenced primitive arrays not returned by _find_all_fs
#186 - Creating subtypes of inheritance-final types (arrays) is not prevented
#187 - The multipleReferencesAllowed flag on array features is not handled

- Consistently wrap arrays into their proper UIMA array types when deserializing
- Tests and code updated because in order to access the array elements it is now necessary to access them via the `elements` property of the array wrapper
- Add handling of the multipleReferencesAllowed flag in during (de)serialization

### CODE REVIEW

File: cassis/cas.py [hunk 1]
```diff
@@ -612,33 +614,52 @@ def _find_all_fs(self, generate_missing_ids: bool = True) -> Iterable[FeatureStr
             all_fs[fs.xmiID] = fs
 
             t = ts.get_type(fs.type)
+
+            # Arrays contents are handled separately - they only have one "virtual" feature: elements
+            if t.supertypeName == "uima.cas.ArrayBase":
+                if t.name == "uima.cas.FSArray":
```
Comment (jcklie) [inline cassis/cas.py:620]: I would swap the condition and do
if t.name != "":
  continue

I hard a hard time to understand the naked continue that is not even in an else block

File: cassis/cas.py [hunk 2]
```diff
@@ -612,33 +614,52 @@ def _find_all_fs(self, generate_missing_ids: bool = True) -> Iterable[FeatureStr
...
+                feature_value = getattr(fs, feature_name)
+                if feature_value is None:
+                    continue
+
                 if (
-                    ts.is_primitive(feature.rangeTypeName)
-                    or ts.is_primitive_collection(feature.rangeTypeName)
-                    or ts.is_primitive_collection(fs.type)
+                    include_inlinable_arrays is False
+                    and feature.multipleReferencesAllowed is not True
```
Comment (jcklie) [inline cassis/cas.py:643]: Why dont you write `not feature.multipleReferencesAllowed`? Similar question to several other locations

File: cassis/typesystem.py [hunk 3]
```diff
@@ -604,10 +612,23 @@ def is_primitive_array(self, type_name) -> bool:
...
             return False
-        elif type_name in _PRIMITIVE_ARRAY_TYPES:
-            return True
-        else:
-            return self.is_primitive_array(self.get_type(type_name).supertypeName)
+
+        # Arrays are inheritance-final, so we do not need to check the inheritance hierarchy
+        return type_name in _PRIMITIVE_ARRAY_TYPES
+
+    def is_array(self, type_name) -> bool:
```
Comment (jcklie) [inline cassis/typesystem.py:619]: type missing

File: cassis/typesystem.py [hunk 4]
```diff
@@ -604,10 +612,23 @@ def is_primitive_array(self, type_name) -> bool:
...
+        """Checks if the type identified by `type_name` is an array.
+
+        Args:
+            type_name: The name of the type to query for.
+        Returns:
+            Returns True if the type identified by `type_name` is an array type, else False
```
Comment (jcklie) [inline cassis/typesystem.py:625]: `True`

File: cassis/xmi.py [hunk 5]
```diff
@@ -290,28 +299,37 @@ def _parse_feature_structure(self, typesystem: TypeSystem, elem, children: Dict[
...
+                feature = AnnotationType.get_feature(feature_name)
+                if typesystem.is_primitive_array(feature.rangeTypeName):
+                    ArrayType = typesystem.get_type(feature.rangeTypeName)
+                    attributes[feature_name] = ArrayType(elements=attributes[feature_name])
+
         self._max_xmi_id = max(attributes["xmiID"], self._max_xmi_id)
         return AnnotationType(**attributes)
 
-    def _parse_primitive_array(self, type_name: str, value: str) -> List:
+    def _parse_primitive_array(self, type: Type, value: str) -> List:
```
Comment (jcklie) [inline cassis/xmi.py:314]: type is a builtin in python, I would call it `type_`

File: cassis/xmi.py [hunk 6]
```diff
@@ -426,35 +469,21 @@ def _serialize_feature_structure(self, cas: Cas, root: etree.Element, fs: Featur
...
-                # </my:fs>
-                for e in value:
+                for e in value.elements:
                     child = etree.SubElement(elem, feature_name)
                     child.text = e
-            elif ts.is_primitive_array(fs.type) and feature_name == "elements":
-                elem.attrib[feature_name] = self._serialize_primitive_array(fs.type, value)
-            elif ts.is_primitive_array(feature.rangeTypeName):
-                elem.attrib[feature_name] = self._serialize_primitive_array(feature.rangeTypeName, value)
+            elif ts.is_primitive_array(feature.rangeTypeName) and feature.multipleReferencesAllowed is not True:
```
Comment (jcklie) [inline cassis/xmi.py:479]: Same question about booleans here

File: tests/test_cas.py [hunk 7]
```diff
@@ -252,9 +252,9 @@ def test_select_returns_feature_structures(cas_with_collections_xmi, typesystem_
     typesystem = load_typesystem(typesystem_with_collections_xml)
     cas = load_cas_from_xmi(cas_with_collections_xmi, typesystem=typesystem)
 
-    arrs = list(cas.select("cassis.StringArray"))
+    arrs = list(cas.select("uima.cas.StringArray"))
```
Comment (jcklie) [inline tests/test_cas.py:255]: We dont need the call to `list` anymore I think, driveby fix pls?

File: cassis/typesystem.py (hunk 3 above)
Comment (jcklie) [reply cassis/typesystem.py:619]: Needs a test

File: cassis/cas.py (hunk 2 above)
Comment (reckart) [reply cassis/cas.py:643]: Because `feature.multipleReferencesAllowed` can be either `False` or `None` or `True` - if it is not `True`, then the condition should kick in. Not sure if it is sensible to write this as `not feature.multipleReferencesAllowed` - I would expect that `None` is not negatable.

File: tests/test_cas.py (hunk 7 above)
Comment (reckart) [reply tests/test_cas.py:255]: Done

File: cassis/typesystem.py (hunk 4 above)
Comment (reckart) [reply cassis/typesystem.py:625]: What do you mean?

File: cassis/cas.py (hunk 1 above)
Comment (reckart) [reply cassis/cas.py:620]: The condition cannot simply be swapped. The continue at the end must always happen, even if the name is not FSArray. I could only swap if I introduced a second continue at the end. IMHO that doesn't improve things.

File: cassis/typesystem.py (hunk 4 above)
Comment (jcklie) [reply cassis/typesystem.py:625]: I use markdown 'True' `True` instead of True for constants and such

File: cassis/cas.py (hunk 2 above)
Comment (jcklie) [reply cassis/cas.py:643]:
```
>>> bool(None)
False
>>> bool(not None)
True

```

File: cassis/cas.py (hunk 1 above)
Comment (jcklie) [reply cassis/cas.py:620]: Ok, that makes sense. It is a bit difficult to read imho, but not so relevant to make it clearer
